### ISSUE TITLE
scons -c fails if .sconsign.dblite is deleted

### ISSUE THREAD

**Issue by bdbaddog** [SCons/scons#2268]
This issue was originally created at: 2008-12-03 12:50:06.
This issue was reported by: `rhl`.
rhl said at 2008-12-03 12:50:06
>scons -c doesn't delete `.sconsign.dblite` and `.sconf_temp` (is this a bug in its own right?)
> 
>Given the problems with a corrupted .scons_temp if you ^C out of a configure context (Issue #2157), I added an `rm -rf .sconsign.dblite .sconf_temp` to the clean actions.
> 
>This results in:
```
scons -c
scons: Reading SConscript files ...
Checking for C++ header file tr1/unordered_map... yes
find . \( -name .svn -prune -o -name \* \) \( -name \*~ -o -name core -o -name
\*.so -o -name \*.os -o -name \*.o \) -exec rm -f {} \; ; rm -rf .sconf_temp
.sconsign.dblite
scons: done reading SConscript files.
scons: Cleaning targets ...
scons: done cleaning targets.
OSError: [Errno 2] No such file or directory: '.sconsign.dblite':
  File "/u/tst/products/DarwinX86/sconsDistrib/1.1.0.d20081104/lib/scons-1.1.0.d20081104/SCons/Script/Main.py", line 1261:
    _exec_main(parser, values)
  File "/u/tst/products/DarwinX86/sconsDistrib/1.1.0.d20081104/lib/scons-1.1.0.d20081104/SCons/Script/Main.py", line 1226:
    _main(parser)
  File "/u/tst/products/DarwinX86/sconsDistrib/1.1.0.d20081104/lib/scons-1.1.0.d20081104/SCons/Script/Main.py", line 989:
    nodes = _build_targets(fs, options, targets, target_top)
  File "/u/tst/products/DarwinX86/sconsDistrib/1.1.0.d20081104/lib/scons-.1.0.d20081104/SCons/Script/Main.py", line 1182:
    jobs.run(postfunc = jobs_postfunc)
  File "/u/tst/products/DarwinX86/sconsDistrib/1.1.0.d20081104/lib/scons-1.1.0.d20081104/SCons/Job.py", line 111:
    postfunc()
  File "/u/tst/products/DarwinX86/sconsDistrib/1.1.0.d20081104/lib/scons-1.1.0.d20081104/SCons/Script/Main.py", line 1179:
    SCons.SConsign.write()
  File "/u/tst/products/DarwinX86/sconsDistrib/1.1.0.d20081104/lib/scons-1.1.0.d20081104/SCons/SConsign.py", line 107:
    syncmethod()
  File "/u/tst/products/DarwinX86/sconsDistrib/1.1.0.d20081104/lib/scons-1.1.0.d20081104/SCons/dblite.py", line 104:
    self._os_unlink(self._file_name)
Exception exceptions.OSError: (2, 'No such file or directory',
'.sconsign.dblite') in <bound method dblite.__del__ of <SCons.dblite.dblite
instance at 0x67fc38>> ignored

```
 
>It is obvously scary to mess with .scon* files, but in this case it
>is needed.

gregnoel said at 2008-12-24 07:13:18
>Bug party triage.  There are plans afoot so that 'scons -ccc' will delete the .sconsign file(s) and the like, but that's still a ways out.  In the meantime, a stack trace is kinda rude, so put this where we'll look at it again.

**Comment by garyemiller** [comment]
This bug is now affecting me.  The .sconsign.dblite made with Python 3 is not compatible with Python 2.  So when I build gpsd with scons on python 3, I need to nuke .sconsign.dblite to do a build on Python 2.

```
scons: Building targets ...
scons: *** [ais_json.os] ValueError : unsupported pickle protocol: 4
```

**Comment by bdbaddog** [comment]
> This bug is now affecting me. The .sconsign.dblite made with Python 3 is not compatible with Python 2. So when I build gpsd with scons on python 3, I need to nuke .sconsign.dblite to do a build on Python 2.
> 
> ```
> scons: Building targets ...
> scons: *** [ais_json.os] ValueError : unsupported pickle protocol: 4
> ```

@garyemiller 
The issue you're having is not this issue, but rather incompatibility between py2 and py3 pickle protocols being used by SCons as well as py2/py3 incompatibilities.

Also this issue is explicitly listed in Announce.txt. (For SCons version 3.0.0)
[https://github.com/SCons/scons/blob/master/src/CHANGES.txt#L412](https://github.com/SCons/scons/blob/master/src/CHANGES.txt#L412
)
Please bring your issue to the users mailing list, IRC, or our discord channel and we'll see if we can help you.

**Comment by garyemiller** [comment]
"The issue you're having is not this issue, but rather incompatibility between py2 and py3 pickle protocols "

This is a cascade of failures.

Yes, of course, we all know what the pickle problem is.

As we all know the solution is simply to delete .sconsign.dblite

But THIS problem is that I can't just fix it ny making "scons -c" delete .sconsign.dblite, since that fails.  Which is THIS bug.

I don't need help, I know how to fix the pickle bug.  scons needs help to fix this 11 year old bug.  

Google .sconsing.dblite.   This bug has hit a lot of people and there is no workaround, except for manual "rm  .sconsign.dblite".

I'm willing to pop onto IRC, if someone tells me what channel.

**Comment by bdbaddog** [comment]
> "The issue you're having is not this issue, but rather incompatibility between py2 and py3 pickle protocols "
> 
> This is a cascade of failures.
> 
> Yes, of course, we all know what the pickle problem is.
> 
> As we all know the solution is simply to delete .sconsign.dblite
> 
> But THIS problem is that I can't just fix it ny making "scons -c" delete .sconsign.dblite, since that fails. Which is THIS bug.
> 
> I don't need help, I know how to fix the pickle bug. scons needs help to fix this 11 year old bug.
> 
> Google .sconsing.dblite. This bug has hit a lot of people and there is no workaround, except for manual "rm .sconsign.dblite".

 
> I'm willing to pop onto IRC, if someone tells me what channel.

#scons on freenode.net of course.
As listed on our website: [https://scons.org/contact.html](https://scons.org/contact.html)

**Comment by mwichmann** [comment]
My understanding is there's really no way to make this work from code in the config scripts; scons has carefully constructed exit code which will cause the file to be rewritten at the very end as the signature database is closed, while instructions you give in SConstruct et.al. would be executed much earlier. Greg Noel's comment about a "really clean" option that has scons itself take care of it would work, but I guess that never came about?

**Comment by bdbaddog** [comment]
> My understanding is there's really no way to make this work from code in the config scripts; scons has carefully constructed exit code which will cause the file to be rewritten at the very end as the signature database is closed, while instructions you give in SConstruct et.al. would be executed much earlier. Greg Noel's comment about a "really clean" option that has scons itself take care of it would work, but I guess that never came about?

Nope. But it does sound like a reasonable idea.  Just make sure the sconsign is removed in an atexit() or similar?

**Comment by mwichmann** [comment]
I'd like to add a doc note somewhere that says not to try to remove the sconsign file in your sconscript, but there's very little talking about them (user guide mentions them in the context of writing a decider, and that's it except for the sconsignfile argument handling), so I'm not seeing a logical place. Any thoughts?

**Comment by mwichmann** [comment]
As to the original bug report, I don't see this happening in a trivial test - manually removing `.sconsign.dblite` after a successful build does not seem to cause any problems in doing `scons -c`. Perhaps in the interim that logic has gotten more robust?

**Comment by garyemiller** [comment]
the .sconsign*dblite files are now removed atexti() when you "scons -c".  So this issue should no longer exist.  If you concur, I'll close this as fixed.

**Comment by mwichmann** [comment]
I'm not seeing the file being removed, nor any code that would do so.

**Comment by garyemiller** [comment]
SConstruct: line 2848:
    atexit.register(lambda: os.system("rm -rf .sconsign*.dblite"))

**Comment by mwichmann** [comment]
I see, in an sconscript. Got it.
